### ISSUE TITLE
Failure to build new checkout on gcc 8.0.1, implied buffer overflow

### ISSUE THREAD

**Issue by PH0lder** [wren-lang/wren#531]
On Fedora 28 Beta, get a fresh copy and do a make, and get warning errors that prevent building.  It seems to imply some sort of possible buffer overflow.

```
make
make[1]: Entering directory '/home/user/Dev/Wren/wren'
       str src/vm/wren_core.wren          
       str src/optional/wren_opt_meta.wren 
       str src/optional/wren_opt_random.wren 
        cc src/optional/wren_opt_meta.c    -O3 -std=c99 -fPIC
        cc src/optional/wren_opt_random.c  -O3 -std=c99 -fPIC
        cc src/vm/wren_utils.c             -O3 -std=c99 -fPIC
        cc src/vm/wren_compiler.c          -O3 -std=c99 -fPIC
src/vm/wren_compiler.c: In function ‘error’:
src/vm/wren_compiler.c:402:36: error: ‘: ’ directive writing 2 bytes into a region of size between 1 and 139 [-Werror=format-overflow=]
   int length = sprintf(message, "%s: ", label);
                                    ^~
src/vm/wren_compiler.c:402:16: note: ‘sprintf’ output between 3 and 141 bytes into a destination of size 139
   int length = sprintf(message, "%s: ", label);
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

**Comment by PH0lder** [comment]
I edited line 402 (as noted above) to add +2 to the constant so that I could continue building.  The next error was:

      run util/build_libuv.py            
        cc src/cli/modules.c               -O3 -std=c99 -fPIC
src/cli/modules.c: In function ‘readBuiltInModule’:
src/cli/modules.c:244:3: error: ‘strncpy’ specified bound depends on the length of the source argument [-Werror=stringop-overflow=]
   strncpy(copy, *module->source, length + 1);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/cli/modules.c:242:19: note: length computed here
   size_t length = strlen(*module->source);
                   ^~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors

As this is in a codebase upon which wren depends, I shall leave it to someone else to report this issue if relevant.  (It sure seems relevant, as, judging from the description, this could be a potential security issue/buffer overflow.)

**Comment by munificent** [comment]
Thanks, I'll take look.

**Comment by munificent** [comment]
OK, I took a stab at this. I can't repro the error locally since I don't have GCC on my Mac, but let me know if this helps or doesn't help.

**Comment by PH0lder** [comment]
I pulled a fresh copy, and did a make, yielding:

```
[user@localhost Wren]$ mv wren wrenOld
[user@localhost Wren]$ git clone https://github.com/munificent/wren.git
Cloning into 'wren'...
remote: Counting objects: 15379, done.
remote: Compressing objects: 100% (44/44), done.
remote: Total 15379 (delta 44), reused 43 (delta 31), pack-reused 15304
Receiving objects: 100% (15379/15379), 6.18 MiB | 762.00 KiB/s, done.
Resolving deltas: 100% (10212/10212), done.
[user@localhost Wren]$ cd wren
[user@localhost wren]$ make
make[1]: Entering directory '/home/user/Dev/Wren/wren'
       str src/vm/wren_core.wren
       str src/optional/wren_opt_meta.wren
       str src/optional/wren_opt_random.wren
        cc src/optional/wren_opt_meta.c    -O3 -std=c99 -fPIC
        cc src/optional/wren_opt_random.c  -O3 -std=c99 -fPIC
        cc src/vm/wren_utils.c             -O3 -std=c99 -fPIC
        cc src/vm/wren_compiler.c          -O3 -std=c99 -fPIC
src/vm/wren_compiler.c: In function ‘error’:
src/vm/wren_compiler.c:402:36: error: ‘: ’ directive writing 2 bytes into a region of size between 1 and 159 [-Werror=format-overflow=]
   int length = sprintf(message, "%s: ", label);
                                    ^~
src/vm/wren_compiler.c:402:16: note: ‘sprintf’ output between 3 and 161 bytes into a destination of size 159
   int length = sprintf(message, "%s: ", label);
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[1]: *** [util/wren.mk:207: build/release/vm/wren_compiler.o] Error 1
make[1]: Leaving directory '/home/user/Dev/Wren/wren'
make: *** [Makefile:16: release] Error 2
```

**Comment by minirop** [comment]
I don't know why, but since the compiler complains about 161 bytes into a 159 buffer, I simply changed this line: `char message[ERROR_MESSAGE_SIZE + 2];` and the error is no more.
(the `+2` corresponds to the colon and the space)

**Comment by PH0lder** [comment]
I have not looked at detail into this.  One fix is probably to use snprintf().  One presumes the issue is that the message variable and label variable are the same length.    I assume the compiler sees that label can be one length and then knows it's adding an additional two chars for the colon and space after it.  The other fix is probably to make sure message has enough room to hold whatever is in label plus two more chars.

**Comment by PH0lder** [comment]
The other issue
  size_t length = strlen(*module->source);
and
  strncpy(copy, *module->source, length + 1);
can be fixed by simply converting this to a strdup(), since that is in essence what is happening anyway.

**Comment by Popolon** [comment]
The problem is still here with gcc 8.2.1

**Comment by vktec** [comment]
This patch fixes it for me:

```diff
diff --git a/src/vm/wren_compiler.c b/src/vm/wren_compiler.c
index adb261f..4be3483 100644
--- a/src/vm/wren_compiler.c
+++ b/src/vm/wren_compiler.c
@@ -398,7 +398,7 @@ static void printError(Parser* parser, int line, const char* label,
   if (parser->vm->config.errorFn == NULL) return;
   
   // Format the label and message.
-  char message[ERROR_MESSAGE_SIZE];
+  char message[ERROR_MESSAGE_SIZE + 2];
   int length = sprintf(message, "%s: ", label);
   length += vsprintf(message + length, format, args);
   ASSERT(length < ERROR_MESSAGE_SIZE, "Error should not exceed buffer.");
```

**Comment by bjorn** [comment]
@vktec While that silences the warning, it needlessly increases the size of the buffer. An alternative fix is provided by #620.
